sdk/opa.go

```diff
@@ -0,0 +1,197 @@
+
+	if opa.paths != nil && len(opa.paths) > 0 {
+		loaded, err := initload.LoadPaths(opa.paths, opa.filter, opa.bundleMode, nil, false)
+		if err != nil {
+			return nil, errors.Wrap(err, "load error")
+		}
+		pluginInitializers = append(pluginInitializers, plugins.InitBundles(loaded.Bundles))
+		pluginInitializers = append(pluginInitializers, plugins.InitFiles(loaded.Files))
+	}
+
+	opa.manager, err = plugins.New(opa.configBytes, id, inmem.New(), pluginInitializers...)
+	if err != nil {
+		return nil, err
+	}
+
+	discovery, err := discovery.New(opa.manager)
+	if err != nil {
+		return nil, err
+	}
+
+	opa.manager.Register("discovery", discovery)
+
+	return opa, nil
+}
+
+// Start asynchronously starts the policy engine's plugins that download
+// policies, report status, etc.
+func (opa *OPA) Start(ctx context.Context) error {
+	return opa.manager.Start(ctx)
+}
+
+// Query sends a query to the OPA instance and returns the decision
+func (opa *OPA) Query(ctx context.Context, decisionPath string, input interface{}, opts ...func(*rego.Rego)) (interface{}, error) {
+	m := metrics.New()
+	var decisionID string
+	var decision interface{}
+
+	err := storage.Txn(ctx, opa.manager.Store, storage.TransactionParams{}, func(txn storage.Transaction) error {
+		var err error
+		decisionID, err = uuid.New(rand.Reader)
+		if err != nil {
+			return err
+		}
+
+		opts = append(opts,
+			rego.Metrics(m),
+			rego.Query(decisionPath),
+			rego.Input(input),
+			rego.Compiler(opa.manager.GetCompiler()),
+			rego.Store(opa.manager.Store),
+			rego.Transaction(txn))
+
+		query, err := rego.New(opts...).PrepareForEval(ctx)
+		if err != nil {
+			return err
+		}
+
+		rs, err := query.Eval(ctx)
+		if err != nil {
+			return err
+		} else if len(rs) == 0 {
+			return fmt.Errorf("undefined decision")
+		} else {
+			decision = rs[0].Expressions[0].Value
+		}
+
+		if logger := logs.Lookup(opa.manager); logger != nil {
+			bundles, err := opa.bundles(ctx, txn)
+
+			record := &server.Info{
+				Bundles:    bundles,
+				DecisionID: decisionID,
+				Timestamp:  time.Now(),
+				Query:      decisionPath,
+				Input:      &input,
+				Error:      err,
+				Metrics:    m,
+			}
+			if err == nil {
+				var x = decision
+				record.Results = &x
+			}
+
+			if err := logger.Log(ctx, record); err != nil {
+				return errors.Wrap(err, "failed to log decision")
+			}
+		}
+
+		return nil
+	})
+
+	return decision, err
+}
+
+func (opa *OPA) bundles(ctx context.Context, txn storage.Transaction) (map[string]server.BundleInfo, error) {
+	bundles := map[string]server.BundleInfo{}
+
+	names, err := bundle.ReadBundleNamesFromStore(ctx, opa.manager.Store, txn)
+	if err != nil && !storage.IsNotFound(err) {
+		panic(err)
+	}
+	for _, name := range names {
+		r, err := bundle.ReadBundleRevisionFromStore(ctx, opa.manager.Store, txn, name)
+		if err != nil && !storage.IsNotFound(err) {
+			panic(err)
+		}
+		bundles[name] = server.BundleInfo{Revision: r}
+	}
+
+	if err != nil {
```

Comment:
Which `err` is that? The only case I can make it out is that it's the one from line 180, and this line could be hit if `storage.IsNotFound(err)` was _true_.